### ISSUE TITLE
Promise long stack traces

### ISSUE THREAD

**Issue by bootstraponline** [puppeteer/puppeteer#979]
Without [promise long stack traces](http://bluebirdjs.com/docs/api/promise.longstacktraces.html), it's impossible to debug failures. Consider this example code:

```javascript
const chrome = require('puppeteer');

(async () => {

let browser = await chrome.launch({headless: false});
let page = await browser.newPage();
page.waitForSelector('a[href="does/not/exist"]', { timeout: 5000 });

})();

```

```bash
> Error: waiting failed: timeout 30000ms exceeded
    at Timeout.WaitTask._timeoutTimer.setTimeout (/project/node_modules/puppeteer/lib/FrameManager.js:453:58)
    at ontimeout (timers.js:469:11)
    at tryOnTimeout (timers.js:304:5)
    at Timer.listOnTimeout (timers.js:264:5)
```

Nowhere in the stack trace is the line number for the code which failed.

**Comment by Janpot** [comment]
It _could_ be done with:
```js
// index.js
global.Promise = require('bluebird');
const puppeteer = require('puppeteer');

(async function () {
  let browser;
  try {
    browser = await puppeteer.launch({headless: false});
    const page = await browser.newPage();
    await page.waitFor('a[href="does/not/exist"]');
  } catch (err) {
    console.log(err.stack);
  } finally {
    if (browser) {
      browser.close();
    }
  }
})();
```

Then, when run as
```sh
BLUEBIRD_DEBUG=1 node .
```

results in

```
Error: waiting failed: timeout 30000ms exceeded
    at Timeout.WaitTask._timeoutTimer.setTimeout [as _onTimeout] (/project/node_modules/puppeteer/lib/FrameManager.js:468:58)
    at ontimeout (timers.js:469:11)
    at tryOnTimeout (timers.js:304:5)
    at Timer.listOnTimeout (timers.js:264:5)
From previous event:
    at new WaitTask (/project/node_modules/puppeteer/lib/FrameManager.js:462:20)
    at Frame.waitForFunction (/project/node_modules/puppeteer/lib/FrameManager.js:413:12)
    at Frame.waitForSelector (/project/node_modules/puppeteer/lib/FrameManager.js:386:17)
    at Frame.waitFor (/project/node_modules/puppeteer/lib/FrameManager.js:369:19)
    at Page.waitFor (/project/node_modules/puppeteer/lib/Page.js:731:29)
    at /project/index.js:9:16
    at <anonymous>
```

That being said. This is not a `puppeteer` issue, this is a [node issue](https://github.com/nodejs/node/issues/11865).

**Comment by bootstraponline** [comment]
`node --inspect` is mentioned in that thread which doesn't seem to work. Bluebird works.

<img width="1032" alt="useless_stack_trace" src="https://user-images.githubusercontent.com/1173057/31317822-e4aac064-ac15-11e7-9322-0684c2ece8e4.png">

Thanks for the link to the node issue. I think it'd be helpful to warn end users of `puppeteer` coming from WebDriver about this. I invested in a small puppeteer automation project which will now have to be rewritten in another language due to node's lack of a sane debugging story for async/await.

**Comment by paulirish** [comment]
Try this again in Node 8.5+ where [async stack traces are introduced](https://github.com/nodejs/node/pull/13870) to node.

I'm using @Janpot's example from above, and you can see the full stack:

![image](https://user-images.githubusercontent.com/39191/31355455-29a20d38-acef-11e7-8d0a-7ed548531126.png)

**Comment by bootstraponline** [comment]
Bluebird works. I wasn't successful in getting the stack traces without bluebird.

**Comment by Janpot** [comment]
@paulirish Yep, I checked that too. But it appeared to me there's a few things left to be desired from that stacktrace. You have blackboxed 12 frames already and still the trace is hard to parse. There's still a bunch of "internal" frames getting in the way. It also looks like you're using an async arrow function instead of the function declaration in my code. This makes you hit a bug in node where it shows `__dirname` instead of `<anonymous>`, not adding to the clarity either. All this issues made me even more impressed with the stacktrace that `bluebird` was able to produce. It's clear and concise and contains just the frames that are needed to debug this code. I've long stopped using `bluebird` in my code but seeing this almost makes me nostalgic 🙂 .
And yeah, I don't want to fire up devtools just to be able to see a reasonable stacktrace.

**Comment by paulirish** [comment]
@ak239 take a look at the last couple comments. Use of `__dirname` as a function name certainly is a bug.  Also this kind of indicates we might want to have some of node's internal scripts blackboxed by default.

**Comment by vsemozhetbyt** [comment]
> Use of __dirname as a function name certainly is a bug

See https://github.com/nodejs/node/issues/15386 and https://bugs.chromium.org/p/v8/issues/detail?id=6822

**Comment by alexkozy** [comment]
There are some ideas about name inferences for arrow functions. Most radical one - adopt Java approach - use outer closure # num as function name. E.g.
`function foo() {
  () => 42 /* foo#1 */
  () => 239 /* foo#2 */
}`

There are two different problems that we need to solve to blackbox internal scripts:
- we should mark internal scripts in more reliable way than `/internal/.*` regexp,
- we should inherit blackboxing state for scripts evaluated from blackboxed one.

@bootstraponline, I just wondering what language and what tool provide nice async experience? Maybe we can borrow some ideas in DevTools.

**Comment by bootstraponline** [comment]
> @bootstraponline, I just wondering what language and what tool provide nice async experience? Maybe we can borrow some ideas in DevTools.

For the project I just finished (automating [bitrise settings with Chrome](https://github.com/bootstraponline/bitrise_settings)) I used Kotlin. I think almost every WebDriver binding other than JavaScript provides synchronous APIs.  async is a terrible fit for UI testing. 

It's sufficiently bad that WebDriver wrote [a promise manager to fake synchronous execution](https://github.com/SeleniumHQ/selenium/issues/2969) for JavaScript. If there's some way for `puppeteer` to solve this at the API level, that'd be amazing. I think due to JavaScript, we're forced into using async.

Personally I like Coroutines from Kotlin for asynchronous programming.
- https://kotlinlang.org/docs/reference/coroutines.html

**Comment by aslushnikov** [comment]
Debugging of async javascript is a separate topic, out of the puppeteer scope. Hopefully, one day we'll have cheap async stacktraces enabled by default in node.js?.. @ak239 🙏

**Comment by cdauth** [comment]
I think fixing this in puppeteer wouldn’t be difficult but would provide a huge benefit to resolving errors.

It can be resolved by creating the Error object _before_ the timeout is reached, so for example in the WaitTask constructor, instead of

```javascript
this._timeoutTimer = setTimeout(() => this.terminate(new Error(`waiting failed: timeout ${timeout}ms exceeded`)), timeout);
```

write it like this:

```javascript
const timeoutError = new Error(`waiting failed: timeout ${timeout}ms exceeded`);
this._timeoutTimer = setTimeout(() => this.terminate(timeoutError), timeout);
```
